tests/functional/test_awsclient.py

```diff
@@ -906,17 +906,48 @@ def test_can_add_permission_for_apigateway_not_needed(self, stubbed_session):
         source_arn = 'arn:aws:execute-api:us-west-2:123:rest-api-id/*'
         policy = {
             'Id': 'default',
-            'Statement': [{
-                'Action': 'lambda:InvokeFunction',
-                'Condition': {
-                    'ArnLike': {
-                        'AWS:SourceArn': source_arn,
-                    }
-                },
-                'Effect': 'Allow',
-                'Principal': {'Service': 'apigateway.amazonaws.com'},
-                'Resource': 'arn:aws:lambda:us-west-2:account_id:function:name',
-                'Sid': 'e4755709-067e-4254-b6ec-e7f9639e6f7b'}],
+            'Statement': [
+                {'Action': 'lambda:NotInvoke',
+                 'Condition': {
+                     'ArnLike': {
+                         'AWS:SourceArn': source_arn,
+                     }
+                 },
+                 'Effect': 'Allow',
+                 'Principal': {'Service': 'apigateway.amazonaws.com'},
+                 'Resource': 'arn:aws:lambda:us-west-2:account_id:function:name',
+                 'Sid': 'e4755709-067e-4254-b6ec-e7f9639e6f7b'},
+                {'Action': 'lambda:InvokeFunction',
+                 'Condition': {
+                     'ArnLike': {
+                         'AWS:SourceArn': 'not-source-arn',
+                     }
+                 },
+                 'Effect': 'Allow',
+                 'Principal': {'Service': 'apigateway.amazonaws.com'},
+                 'Resource': 'arn:aws:lambda:us-west-2:account_id:function:name',
+                 'Sid': 'e4755709-067e-4254-b6ec-e7f9639e6f7b'},
+                {'Action': 'lambda:InvokeFunction',
+                 'Condition': {
+                     'ArnLike': {
+                         'AWS:SourceArn': source_arn,
+                     }
+                 },
+                 'Effect': 'Allow',
```

Comment:
Without looking at what got removed in the diff, it is not very easy to tell what portion of the policy needs to be present for it to be skipped. Maybe separate this into tests where one tests that a specific policy entry skips over the `add_permission` call but the other entries causes `add_permission` to be triggered?